jaxopt/_src/lbfgs.py

```diff
@@ -0,0 +1,312 @@
+    Algorithm 7.4 (page 178).
+  """
+  fun = partial(inv_hessian_product_leaf, rho_history=rho_history, gamma=gamma)
+  return tree_map(fun, pytree, s_history, y_history)
+
+
+def compute_gamma(s_history: Any, y_history: Any):
+  # Let gamma = vdot(y_history[-1], s_history[-1]) / sqnorm(y_history[-1]).
+  # The initial inverse Hessian approximation can be set to gamma * I.
+  # See Numerical Optimization, second edition, equation (7.20).
+  # Note that unlike BFGS, the initialization can change on every iteration.
+
+  fun = lambda s_history, y_history: tree_vdot(y_history[-1], s_history[-1])
+  num = tree_sum(tree_map(fun, s_history, y_history))
+
+  fun = lambda y_history: tree_vdot(y_history[-1], y_history[-1])
+  denom = tree_sum(tree_map(fun, y_history))
+
+  return jnp.where(denom > 0, num / denom, 1.0)
+
+
+def init_history(pytree, history_size):
+  fun = lambda leaf: jnp.zeros((history_size,) + leaf.shape)
+  return tree_map(fun, pytree)
+
+
+def _update_history(history_array, new_value):
+  """Shift past elements to the left and add new elements at the end."""
+  # TODO: to avoid memory copies, it would be more efficient to treat history as
+  # a rolling buffer, and only set a single vector.
+  return jnp.roll(history_array, -1, axis=0).at[-1].set(new_value)
+
+
+def update_history(history_pytree, new_pytree):
+  return tree_map(_update_history, history_pytree, new_pytree)
+
+
+class LbfgsState(NamedTuple):
+  """Named tuple containing state information."""
+  iter_num: int
+  stepsize: float
+  error: float
+  s_history: Any
+  y_history: Any
+  rho_history: jnp.ndarray
+  aux: Optional[Any] = None
+
+
+@dataclass(eq=False)
+class LBFGS(base.IterativeSolver):
+  """LBFGS solver.
+
+  Attributes:
+    fun: a smooth function of the form ``fun(x, *args, **kwargs)``.
+    has_aux: whether function fun outputs one (False) or more values (True).
+      When True it will be assumed by default that fun(...)[0] is the objective.
+
+    maxiter: maximum number of proximal gradient descent iterations.
+    tol: tolerance of the stopping criterion.
+
+    maxls: maximum number of iterations to use in the line search.
+    decrease_factor: factor by which to decrease the stepsize during line search
+      (default: 0.8).
+    increase_factor: factor by which to increase the stepsize during line search
+      (default: 1.5).
+
+    history_size: size of the memory to use.
+    use_gamma: whether to initialize the inverse Hessian approximation with
+      gamma * I, see 'Numerical Optimization', equation (7.20).
+
+    implicit_diff: whether to enable implicit diff or autodiff of unrolled
+      iterations.
+    implicit_diff_solve: the linear system solver to use.
+
+    jit: whether to JIT-compile the optimization loop (default: "auto").
+    unroll: whether to unroll the optimization loop (default: "auto").
+
+    verbose: whether to print error on every iteration or not.
+      Warning: verbose=True will automatically disable jit.
```

Comment:
verbose is an int, but in the doc it's treated as a bool. perhaps we should say `Any value of verbose greater than zero will automatically disable ...`?

Reply:
Ok switched to boolean for now as I don't need a verbosity level at the moment.